components/integrations/bots/bots.jsx

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import {FormattedMessage} from 'react-intl';
+
+import * as Utils from 'utils/utils.jsx';
+import BackstageList from 'components/backstage/components/backstage_list.jsx';
+import Constants from 'utils/constants.jsx';
+
+import Bot from './bot.jsx';
+
+export default class Bots extends React.PureComponent {
+    static propTypes = {
+
+        /**
+        *  Map from botUserId to bot.
+        */
+        bots: PropTypes.object.isRequired,
+
+        /**
+        *  Map from botUserId to accessTokens.
+        */
+        accessTokens: PropTypes.object.isRequired,
+
+        /**
+        *  Map from botUserId to owner.
+        */
+        owners: PropTypes.object.isRequired,
+
+        actions: PropTypes.shape({
+
+            /**
+            * Ensure we have bot accounts
+            */
+            loadBots: PropTypes.func.isRequired,
+
+            /**
+            * Load access tokens for bot accounts
+            */
+            getUserAccessTokensForUser: PropTypes.func.isRequired,
+
+            /**
+            * Access token managment
+            */
+            createUserAccessToken: PropTypes.func.isRequired,
+            revokeUserAccessToken: PropTypes.func.isRequired,
+            enableUserAccessToken: PropTypes.func.isRequired,
+            disableUserAccessToken: PropTypes.func.isRequired,
+
+            /**
+            * Load owner of bot account
+            */
+            getUser: PropTypes.func.isRequired,
+
+            /**
+            * Disable a bot
+            */
+            disableBot: PropTypes.func.isRequired,
+
+            /**
+            * Enable a bot
+            */
+            enableBot: PropTypes.func.isRequired,
+        }),
+
+        /**
+        *  Only used for routing since backstage is team based.
+        */
+        team: PropTypes.object.isRequired,
+    }
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            loading: true,
+        };
+    }
+
+    componentDidMount() {
+        this.props.actions.loadBots(
+            Constants.Integrations.START_PAGE_NUM,
+            Constants.Integrations.PAGE_SIZE
+        ).then(
+            (result) => {
+                if (result) {
+                    // We don't need to wait for this and we need to accept failure in the case where bot.owner_id is a plugin id
+                    result.data.map((bot) => this.props.actions.getUser(bot.owner_id));
+                    Promise.all(
```

Comment:
I'm having trouble following this because of the nesting and indentation. Maybe we could turn this into an explicit loop to make it easier to follow?

```javascript
const promises = [];

for (const bot of result.data) {
    promises.push(this.props.actions.getUser(bot.user_id));
    promises.push(this.props.actions.getUserAccessTokensForUser(bot.user_id));
}

Promise.all(promises).then(() => {
    this.setState({loading: false});
});
```

Reply:
Ya. This steadily grew as I added features. I agree with your cleanup suggestion.